Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -3524,21 +3524,29 @@ private async Task<PipExecutionStep> ExecutePipStep(RunnablePip runnablePip)
             }
         }
 
-        private static void FlagSharedOpaqueOutputs(IPipExecutionEnvironment environment, ExecutionResult executionResult)
+        private void FlagSharedOpaqueOutputs(IPipExecutionEnvironment environment, ProcessRunnablePip process)
         {
+            // Select all declared output files
+            IEnumerable<(AbsolutePath, bool)> paths = process.Process
+                .FileOutputs
+                .Select(f => (f.Path, false));

```

Comment:
Maybe I'm missing something, but why do you need to do this for files, if you are already calling MakeSharedOpaqueOutputIfNeeded under ReportContent for every file being reported?

Reply:
I'm not 100% sure this is needed either.

My understanding is that this code path exists so that it flags files even if the pip failed, which is why I thought I'd include all outputs.

When the pip doesn't fail, this code path is indeed not necessary

Reply:
From what I can see, if you leave this code as is (which seems fine, in the sense that even when pips are terminated early, this will run, since that's working for explicit shared opaque content), then you don't need the change you added in ReportContent, since that step will always re-do work.

Reply:
That's true for the case when the pip runs and produces the files; when the pip is replayed from cache, then only `ReportContent` gets called.

Reply:
As far as I remember I also modified how the file materialization works to ensure timestamps there as well

Reply:
You did, but I removed it in this PR.

The reason I removed it is because there I don't have access to PipGraph or something that would let me ask if a give path is under any output directory.  I could add a method to `IFileContentManagerHost` that would allow me to do this though

Reply:
I think one other reason to keep the timestamp enforcement during materialization is: a file that is not under a shared opaque is cached, then there is an unrelated new pip defining a shared opaque that includes the file. The first pip is replayed from cache. We need to make sure the timestamp is now changed.
Not sure if all this happens when reporting content. Not 100$ sure, but if I remember correctly, it doesn't.